openshift/odo
# move URL info to devfile

**What type of PR is this?**

/area devfile
/area url



**What does does this PR do / why we need it**:
This PR moves most of URL info from env.yaml to devfile, for Route and Ingress URL type for devfile components

Upon url creation, it will save the URL info into devfile endpoints. Only `urlName`, `urlKind`, `host` and `tlsSecret`
(for ingress type only) will be saved in env.yaml. devfile endpoint will contain information of `Name`(should be as urlName), `TargetPort`, `Exposure`, `Secure`, `Protocol`, `Path`. 
created `--path`, `--exposure`, `--protocol` flags for `odo url create` to collect those specific settings. 
The settings will be validated. If not provided, default values will be assigned according to definition in devfile schema: https://devfile.github.io/devfile/_attachments/api-reference.html

 
url list and describe checks and treats devfile endpoints as local URLs
url delete would remove the specific endpoint with the provided name from devfile 


**Which issue(s) this PR fixes**:

Fixes #3718 
Fixes #3615 

**PR acceptance criteria**:




- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

**Automatim tests:**
run `make test-cmd-devfile-url` for integration tests and `make test` for unit tests

**Manual tests:**
`odo create java-springboot`
`odo url create url1 --port <portNumber> <any arguments you want to specify>`
should see endpoint with `url1` created in devfile.yaml, with all information provided within `odo url create`

`odo url describe url1` should see url1's info, and state is `not pushed`

`odo push` should see message that url1 is getting created. 

`odo url describe url1` should see url1's info, and state is `pushed`

`odo url list` should see url1 and the original url `8080/tcp` defined in springoot devfile.

`odo url delete url1 ` should successfully delete the url info saved in devfile.yaml and env.yaml
